vm/devices/support/fs/lxutil/src/windows/readdir.rs

```diff
@@ -0,0 +1,651 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+use super::fs;
+use super::util;
+use crate::windows::path;
+use bitfield_struct::bitfield;
+use pal::windows::UnicodeString;
+use std::ffi;
+use std::os::windows::io::AsRawHandle;
+use std::os::windows::io::FromRawHandle;
+use std::os::windows::io::OwnedHandle;
+use std::ptr;
+use windows::Wdk::Storage::FileSystem;
+use windows::Wdk::Storage::FileSystem::FILE_INFORMATION_CLASS;
+use windows::Wdk::System::Threading;
+use windows::Win32::Foundation;
+use windows::Win32::Storage::FileSystem as W32Fs;
+use windows::Win32::System::Kernel;
+use windows::Win32::System::Memory;
+use windows::Win32::System::Threading as W32Threading;
+
+const DIR_ENUM_BUFFER_SIZE: usize = 4096;
+const BUFFER_EXTRA_SIZE: usize = 0x200;
+
+#[bitfield(u32)]
+struct DirectoryEnumeratorFlags {
+    end_reached: bool,
+    asynchronous_mode: bool,
+    #[bits(30)]
+    _reserved: u32,
+}
+
+// Some of these FileInformationClasses are missing from windows-rs.
+#[derive(PartialEq)]
+enum DirectoryEnumeratorFileInformationClass {
+    FileId64ExtdDirectoryInformation,
+    FileIdAllExtdDirectoryInformation,
+    FileIdExtdDirectoryInformation,
+    FileIdFullDirectoryInformation,
+    FileIdBothDirectoryInformation,
+    FileFullDirectoryInformation,
+    FileDirectoryInformation,
+}
+
+#[allow(non_camel_case_types, non_snake_case, unused)]
+#[repr(C)]
+struct FILE_ID_64_EXTD_DIR_INFORMATION {
```

Comment:
You might consider putting all your ABI types in a single module on their own, rather than have them scattered about along with code.